### DISCUSSION TITLE
feature: Direction offset in animation editor and preview on sprite

### DISCUSSION THREAD

**Discussion by pandinocoder** [AscensionGameDev/Intersect-Engine#2714] [Ideas]
We want to set animation sprites per direction, based on the direction the player is facing, using the following mockup:

![image](https://github.com/AscensionGameDev/Intersect-Engine/assets/26048926/c898ed21-676a-42c2-b243-12ce4de3265f)

Green outline: where the user can manage these offsets. The list is auto-populated with the directions a player can look in (4 or 8 directions, depending on the server settings) 

Red outline: the original Revert Changes button which is now made smaller. Add a tooltip that says: Revert changes

- The dropdown should be grouped and then sorted using the current culture (**not invariant or ordinal**):
  1. Names that contain the original animation name will be grouped first (e.g. `fireball_up` contains `fireball`, but `fire` is contained by `fireball`)
     - Names in this group should be additionally subgrouped by the difference in number of underscores from the original animation name (e.g. `fireball_up` is sorted before `fireball_down_right` because `fireball_up` has 1 underscore and is considered "more similar" to the original name `fireball`)
  2. Names that are contained by the original animation name will be grouped second
     - Names should be additionally subgrouped by their `IndexOf()` value in the original name (e.g. `fire` is sorted before `ball` because `fire` is at index 0 in `fireball` while `ball` is at index 4, and `reball` would be sorted before `ball` as well because its index would be 2). 
  3. All other names, sorted solely according to the current culture, no subgroups

---

<sup>Partial text from the original suggestion posted by **da-rulez** November  4, 2023:</sup>
Right now the animation system works perfectly fine for animations that play directly on the player, but making animations for spells that need to rotate and aren't centered is often very hard, and sometimes impossible. 

I propose that we make this easier, and at the same time make troubleshooting any animation easier as well: 

New option to display an entity in the animation preview box, so you can see how it will look like on that entity (obviously this would need to render in the same way as it does ingame) -> Preview Sprite button is where you can set a graphic.

This selection works similar to setting an event graphic, you can select any frame from any entity animation: 

![image](https://github.com/AscensionGameDev/Intersect-Engine/assets/26048926/e0d305d0-5a8f-4424-aecf-061a70811e87)

**Comment by dhmello** [comment]
Are these options per frames or are they global?

It would be great if you could configure it by frames. If you already have this idea, great.

**Comment by da-rulez** [comment]
> Are these options per frames or are they global?
> 
> It would be great if you could configure it by frames. If you already have this idea, great.

Global, I don't really see in which case a frame by frame offset couldn't be solved in the animation graphic itself. Let me know

**Comment by pandinocoder** [comment]
@theceltos Can you provide a common use case where it is valuable per frame?

Each frame should already be aligned where it is supposed to be relative to the previous frame, I do not see how a per-frame offset is going to be valuable.

**Comment by dhmello** [comment]
I imagine a fireball in the character's hand, where the sprite will have a throwing movement for example, the animation being able to be followed over the character's hand is just an idea.

Or even being able to specify from which angle it can be animated, imagine a wing for example, being able to make an animation of the character going up, X animation, going right, Y animation, etc., would be a great implement in this function

**Comment by pandinocoder** [comment]
I think the fireball example is a decent idea, but I also think it goes far beyond the effort most people would be putting in.

There is also no space left in the UI to add frame-specific offsets so I think you should open a discussion referencing this issue to expand upon the feature to add frame-specific offsets, but this will not be addressed until we finish rewriting the editor.

**Comment by dhmello** [comment]
A quick solution for animations based on the character's directions would be to place this in the spell editor, where when the spell is used, it will trigger the animation corresponding to the direction the character is looking.

However, this change is already a big leap for the animation editor, and I'm excited to be able to use it one day ^.^

**Comment by Ezekyuhl** [comment]
Kas and I were talking about an issue I was having that may relate to this.

Issue: 
When rotations are disabled, the moth is always in the same spot (almost okay, but the Left direction makes it look silly)
When rotations are enabled the moth rotates and pops to all different orientations.
Goal: I would like it to behave as the spritesheet should.

Video:
https://github.com/AscensionGameDev/Intersect-Engine/assets/17024553/1f1296df-dbac-494d-8529-723eb36fca5e

**Idea being that we can set animations per direction:**

![image (6)](https://github.com/AscensionGameDev/Intersect-Engine/assets/17024553/84a7eb69-8157-4de9-9440-db231b69780d)

(In my case, I think it would also work if a guy could add equipment to the AnimatedSprites list in config.json too, but if this gives more overall use it would be better.)

**Comment by Ezekyuhl** [comment]
Another thing I was attempting to do where I ran into the need to **_set animation by orientation_** again was emotes, I was going to overlay the facial expressions onto the sprite as an animation, it won't work that way.

**Comment by da-rulez** [comment]
The goal is to have an animation per direction instead: 

![image](https://github.com/AscensionGameDev/Intersect-Engine/assets/26048926/6d6fff85-77cf-47b3-aba1-5c32e90c5dff)
![image](https://github.com/AscensionGameDev/Intersect-Engine/assets/26048926/3b694def-04a1-4253-bb64-ce473ff2d306)
